### ISSUE TITLE
please develop functions of verilog

### ISSUE THREAD

**Issue by Geek-wireless** [doxygen/doxygen#6574]

**Comment by albert-github** [comment]
verilrog is, to the best of my knowledge a HDL (hardware description language). VHDL is also a HDL so:

- what are the differences between verilog and VHDL
- what is missing for verilog in the VDHL implementation in doxygen
- where to obtain a (free) standard of the verilog standard / language

**Comment by norwarl** [comment]
Syntactically verilog is much closer to C than VHDL.  

If a verilog implementation is started it would be a good idea to start supporting SystemVerilog instead/as well as that is more feature rich and would have much greater benefit of Doxygen support since it supports classes and similar constructs used in verification. 

Someone made a perl filter for SystemVerilog to turn the files into something C-like https://github.com/SeanOBoyle/DoxygenFilterSystemVerilog

There is also a verilog fork of Doxygen https://github.com/avelure/doxygen-verilog 

SystemVerilog LRM can be had from IEEE https://ieeexplore.ieee.org/document/8299595
It's free, but you need an account to access it.

**Comment by rw1nkler** [comment]
Hello, is there any interest in adding Verilog support to Doxygen? What needs to be done to provide this kind of functionality? Is there any information about the process of adding a new programming language to Doxygen?

**Comment by mpkopec** [comment]
@rw1nkler Yes, there is a need, as there is no tool I could find to create automated documentation for Verilog. The solution from this repository: https://github.com/avelure/doxygen-verilog works quite well, but the version of Doxygen is old and it does not seem to be maintained.

**Comment by tobknight** [comment]
> @rw1nkler Yes, there is a need, as there is no tool I could find to create automated documentation for Verilog. The solution from this repository: https://github.com/avelure/doxygen-verilog works quite well, but the version of Doxygen is old and it does not seem to be maintained.

Can I know current status looks like?  So that repo is not usble any more or? I would like to know more detail

**Comment by mpkopec** [comment]
@tobknight The repo from @avelure is usable, however the relations in my opinion are presented in reverse, there is also still a need for elaboration of the modules as well as a better IO documenting options.

**Comment by albert-github** [comment]
Well it looks like that repository hasn't been updated for a while (as the fork says at the moment: "This branch is 11 commits ahead, 3318 commits behind doxygen:master. ") and thus a lot of updates and improvements in doxygen are not included. Also from the owner (@avelure) the are no activities on github at all.
So I think the usability is limited.

Points that should be investigated:
- what has been changed in the forked repository (i.e. comparing against the official doxygen version at that moment) and based on the changes see whether it is possible to add these changes to the official repository by means of a proposed pull request (I cannot guarantee that it will be integrated, depends on the changes and impact).
- @avelure do you have any intentions to update your repository and / or make a PR for doxygen?

**Comment by avelure** [comment]
My fork https://github.com/avelure/doxygen-verilog is based of the earlier DoxVerilog https://bitbucket.org/maddoxx/doxverilog/src/master/ which piggybacks on the VHDL implementation in Doxygen and uses verilog-perl https://github.com/veripool/verilog-perl as a parser.
The work is quite hacky in nature and does not lend it self to a clean PR in the present state I believe. Among other things, it is not possible to have both Verilog and VHDL in the same documentation and there are no tests.
My work consist mostly of porting it to newer Doxygen release then the original 1.8.5

There was an effort from @AIS-L to move it to more recent Doxygen version, as well as add SystemVerilog support, but it seems to have stalled as well.

You can have a look at the SystemVerilog-dev branch for the most recent version. I might have uncommitted work-in-progress for a later doxygen version, I will see what I can do in the comming weeks.

I unfortunately don't have much time to work on this, but I'm open for help if anyone wants to contribute.

**Comment by albert-github** [comment]
@avelure 
> You can have a look at the SystemVerilog-dev branch for the most recent version.

Where can I find this branch?


@maddox11 what is your opinion on all of this?

**Comment by avelure** [comment]
https://github.com/avelure/doxygen-verilog/tree/SystemVerilog-dev

**Comment by tobknight** [comment]
> My fork https://github.com/avelure/doxygen-verilog is based of the earlier DoxVerilog https://bitbucket.org/maddoxx/doxverilog/src/master/ which piggybacks on the VHDL implementation in Doxygen and uses verilog-perl https://github.com/veripool/verilog-perl as a parser.
> The work is quite hacky in nature and does not lend it self to a clean PR in the present state I believe. Among other things, it is not possible to have both Verilog and VHDL in the same documentation and there are no tests.
> My work consist mostly of porting it to newer Doxygen release then the original 1.8.5
> 
> There was an effort from @AIS-L to move it to more recent Doxygen version, as well as add SystemVerilog support, but it seems to have stalled as well.
> 
> You can have a look at the SystemVerilog-dev branch for the most recent version. I might have uncommitted work-in-progress for a later doxygen version, I will see what I can do in the comming weeks.
> 
> I unfortunately don't have much time to work on this, but I'm open for help if anyone wants to contribute.

What if we integrate a System-Verilog Parser to handle its documentation exclusive? Will this resolve the documentation on these two languages?

**Comment by avelure** [comment]
verilog-perl also parses SystemVerilog, the branch is more related to doxygen things I believe, like it doesn't pick up .sv and .svh files. I'm not sure what changing to another parser would resolve.

**Comment by tobknight** [comment]
> verilog-perl also parses SystemVerilog, the branch is more related to doxygen things I believe, like it doesn't pick up .sv and .svh files. I'm not sure what changing to another parser would resolve.

Many thanks

**Comment by albert-github** [comment]
@avelure  you wrote:
> verilog-perl also parses SystemVerilog, the branch is more related to doxygen things I believe, like it doesn't pick up .sv and .svh files. I'm not sure what changing to another parser would resolve.

do you see this as a problem of `verilog-perl` or as a doxygen problem ? In the later case did you look at the `EXTENSION_MAPPING` setting?

**Comment by avelure** [comment]
I believe `verilog-perl` has sufficient SystemVerilog implementation, though it is probably not complete and not as actively developed anymore, other parsers claiming full compatibility exists. In this case I think the changes were related more to that the features of the language is not taken into account in the modifications that were done to Doxygen, the extensions is a minor change of course. 

SystemVerilog is an evolution of Verilog containing object-oriented programming and higher level functionality. You can see it as going from ANSI C to C++V14, it needs some slight adjustments for display purposes.

**Comment by omasanori** [comment]
For someone who wants to implement this, https://chipsalliance.github.io/sv-tests-results/ provides a list of free and open-source SystemVerilog implementations including parsers. IMHO Surelog or Verible may be suitable for integration as they are written in C++.

**Comment by AWHoppe** [comment]
Has support for SystemVerilog been developed further or has this issue died?

**Comment by Dragonfire3900** [comment]
I'm very interested in trying to get [the mentioned fork](https://github.com/avelure/doxygen-verilog) in a mergable state. 

Other than updating to the newest version of doxygen are there other requirements?

**Comment by omasanori** [comment]
Maybe some example and test files in the corresponding directories?

**Comment by maddox11** [comment]
I have developed a SystemVerilog parser ( not completed yet)(see https://github.com/maddox11/doxygen_antlr) 
Antlr serves as parser .
But I don't know if the Doxygen license is compatible with those of the Antlr library ??

**Comment by maddox11** [comment]
complete the SystemVerilog parser see https://github.com/maddox11/sysverilog_antlr4_parser
